game-core/src/main/java/games/strategy/engine/data/unit/ability/ConvertUnitAbility.java

```diff
@@ -0,0 +1,64 @@
+package games.strategy.engine.data.unit.ability;
+
+import games.strategy.engine.data.UnitType;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.experimental.FieldDefaults;
+import lombok.experimental.NonFinal;
+
+/** Unit Ability to convert CombatUnitAbilities into other CombatUnitAbilities or into nothing */
```

Comment:
Expounding, this topic does make me wonder if the concept of unit abilities is combat specific, generic, and how that relates to support. Are those three concepts, two, or just one concept?

For example, could we one day be asked to support an ability that makes units no-longer air (and hence targetable by say land units that could not otherwise target air?).

Reply:
I didn't go with `AntiUnitAbility` as I explained in the forum.  I need to be able to convert unit abilities from one to another and have that be affected by both units on the same side and on the opposite side.  `Anti` didn't make sense in that regard.

Here's some examples:
* An air unit can not hit a sub unless a destroyer is present.  So the air unit initially has an ability to target all units except the subs.  The destroyer gets a `ConvertUnitAbility` that will swap out the air unit`s normal ability with an ability that allows it to target the subs.  This is a friendly destroyer helping an air unit.
* A sub gets a first strike unless a destroyer is present.  So the sub initially has an ability to target during the first strike phase.  The destroyer gets a `ConvertUnitAbility` that will swap out the subs normal ability with an ability that makes the sub fire during the general phase.
* An AA unit has a `willNotFireIfPresent` attribute that is set to `Tank`.  So if a tank is present, the AA won't fire.  So the AA initially has an ability to fire during the AA phase.  The tank gets a `ConvertUnitAbility` that will remove the AA's ability so that it doesn't do anything.

> So bottom line, it's not fully clear to me how many unit ability concepts we have and which ones would fall into which categories. I was hoping we could spend some time discussing and thinking this through so we have our mental and object models more coherent and thought out.

Do you want this discussion in the forum?  Do you want to start it?

> Expounding, this topic does make me wonder if the concept of unit abilities is combat specific, generic, and how that relates to support. Are those three concepts, two, or just one concept?

I don't know how to answer that since I don't know how you are conceptualizing those concepts.  It looks like you might be envisioning these "unit abilities" in a much larger context than I'm envisioning.

> For example, could we one day be asked to support an ability that makes units no-longer air (and hence targetable by say land units that could not otherwise target air?).

This is already possible using AA `targetsAA` or non-AA `canNotBeTargetedBy` and `canNotTarget`.  This would just unify how those are handled.

Reply:
> Do you want this discussion in the forum? Do you want to start it?

I'm not a fan of how we have discussions split between forum and git. So far it's basically the more code focused topics are in git, and conversations that involve map makers and more are in forums (in part because the latter only check forums). IMO it does not matter too much where, I do think the conversation is more code focused as it is more about modelling, how the concepts are modeled out to be mostly transparent. To the extent that model is exposed in XML, that is where the larger community would want to be involved for sure.

I don't quite have enough context and a vision of where these updates are going to start that conversation. Said another way, this is a place where I think good up-front design will be beneficial.

Reply:
My vision is really just to unify how units target.  All of the pieces that I've modeled in the unit abilities (and the old firing squadron pr) is only focused on that.  I want to simplify the battle logic instead of having 4 special hardcoded fight phases, instead it can just loop through a list of phases and ask each phase how to handle the combat.

So support and other abilities (such as blitz) are outside the scope in my vision.  Could they be merged into one unified idea?  Maybe.  Same with retreat, withdraw, and paratroop landing (which are part of a battle but not related with targeting).